### ISSUE TITLE
Connections to CQL binary port leaked when calling cql3SaveToCassandra

### ISSUE THREAD

**Issue by brenttheisen** [tuplejump/calliope#23]
I have a Spark job that is unable to complete successfully when it attempts to persist a large RDD to a CQL table using `cql3SaveToCassandra`. The error I'm seeing in the app log for the job is...

```
SEVERE: Exception in task ID 996
java.lang.RuntimeException
        at org.apache.cassandra.hadoop.cql3.CqlPagingRecordReader$RowIterator.executeQuery(CqlPagingRecordReader.java:642)
        at org.apache.cassandra.hadoop.cql3.CqlPagingRecordReader$RowIterator.computeNext(CqlPagingRecordReader.java:318)
        at org.apache.cassandra.hadoop.cql3.CqlPagingRecordReader$RowIterator.computeNext(CqlPagingRecordReader.java:285)
        at com.google.common.collect.AbstractIterator.tryToComputeNext(AbstractIterator.java:143)
        at com.google.common.collect.AbstractIterator.hasNext(AbstractIterator.java:138)
        at org.apache.cassandra.hadoop.cql3.CqlPagingRecordReader.nextKeyValue(CqlPagingRecordReader.java:201)
        at com.tuplejump.calliope.cql3.Cql3CassandraRDD$$anon$1.hasNext(Cql3CassandraRDD.scala:73)
        at scala.collection.Iterator$$anon$11.hasNext(Iterator.scala:327)
        at scala.collection.Iterator$class.foreach(Iterator.scala:727)
        at scala.collection.AbstractIterator.foreach(Iterator.scala:1157)
        at scala.collection.generic.Growable$class.$plus$plus$eq(Growable.scala:48)
        at scala.collection.mutable.ArrayBuffer.$plus$plus$eq(ArrayBuffer.scala:103)
        at org.apache.spark.CacheManager.getOrCompute(CacheManager.scala:75)
        at org.apache.spark.rdd.RDD.iterator(RDD.scala:230)
        at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:161)
        at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:102)
        at org.apache.spark.scheduler.Task.run(Task.scala:53)
        at org.apache.spark.executor.Executor$TaskRunner$$anonfun$run$1.apply$mcV$sp(Executor.scala:213)
        at org.apache.spark.deploy.SparkHadoopUtil.runAsUser(SparkHadoopUtil.scala:49)
        at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:178)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
        at java.lang.Thread.run(Thread.java:744)
Caused by: TimedOutException()
        at org.apache.cassandra.thrift.Cassandra$execute_prepared_cql3_query_result$execute_prepared_cql3_query_resultStandardScheme.read(Cassandra.java:53671)
        at org.apache.cassandra.thrift.Cassandra$execute_prepared_cql3_query_result$execute_prepared_cql3_query_resultStandardScheme.read(Cassandra.java:53630)
        at org.apache.cassandra.thrift.Cassandra$execute_prepared_cql3_query_result.read(Cassandra.java:53545)
        at org.apache.thrift.TServiceClient.receiveBase(TServiceClient.java:78)
        at org.apache.cassandra.thrift.Cassandra$Client.recv_execute_prepared_cql3_query(Cassandra.java:1820)
        at org.apache.cassandra.thrift.Cassandra$Client.execute_prepared_cql3_query(Cassandra.java:1805)
        at org.apache.cassandra.hadoop.cql3.CqlPagingRecordReader$RowIterator.executeQuery(CqlPagingRecordReader.java:631)
        ... 22 more
```

At roughly the same time my Cassandra log has this entry...

```
ERROR [Thread-4] 2014-05-05 08:23:38,310 CassandraDaemon.java (line 192) Exception in thread Thread[Thread-4,5,main]
java.lang.OutOfMemoryError: unable to create new native thread
        at java.lang.Thread.start0(Native Method)
        at java.lang.Thread.start(Thread.java:713)
        at java.util.concurrent.ThreadPoolExecutor.addWorker(ThreadPoolExecutor.java:949)
        at java.util.concurrent.ThreadPoolExecutor.execute(ThreadPoolExecutor.java:1371)
        at org.apache.cassandra.thrift.CustomTThreadPoolServer.serve(CustomTThreadPoolServer.java:108)
        at org.apache.cassandra.thrift.ThriftServer$ThriftServerThread.run(ThriftServer.java:136)
```

Research of this Cassandra `OutOfMemoryError` told me it is most likely being caused by hitting the open file descriptor limit set in limits.conf. Upon further examination with `netstat` and `sysctl net.netfilter.nf_conntrack_count` I was able to find that the Spark executor process was establishing over 65,000 connections (my `sysctl net.netfilter.nf_conntrack_count` limit is 65,536) to the Cassandra's CQL binary port 9160. Here is a very small excerpt from running `netstat` while the executor is attempting `cql3SaveToCassandra`...

```
tcp        0      0 172.31.37.1:9160            172.31.37.2:46760           ESTABLISHED 5216/java
tcp        0      0 172.31.37.1:9160            172.31.37.2:35263           ESTABLISHED 5216/java
tcp        0      0 172.31.37.1:9160            172.31.37.1:58011           ESTABLISHED 5216/java
tcp        0      0 172.31.37.1:9160            172.31.37.2:49786           ESTABLISHED 5216/java
tcp        0      0 172.31.37.1:9160            172.31.37.3:35322           ESTABLISHED 5216/java
tcp        0      0 172.31.37.1:9160            172.31.37.3:46685           ESTABLISHED 5216/java
tcp        0      0 172.31.37.1:9160            172.31.37.1:47116           ESTABLISHED 5216/java
tcp        0      0 172.31.37.1:9160            172.31.37.3:40039           ESTABLISHED 5216/java
tcp        0      0 ::ffff:172.31.37.1:56924    ::ffff:172.31.37.1:9160     ESTABLISHED 8520/java
tcp        0      0 ::ffff:172.31.37.1:51757    ::ffff:172.31.37.3:9160     ESTABLISHED 8520/java
tcp        0      0 ::ffff:172.31.37.1:54380    ::ffff:172.31.37.1:9160     ESTABLISHED 8520/java
tcp        0      0 ::ffff:172.31.37.1:53491    ::ffff:172.31.37.1:9160     ESTABLISHED 8520/java
tcp        0      0 ::ffff:172.31.37.1:46919    ::ffff:172.31.37.3:9160     ESTABLISHED 8520/java
tcp        0      0 ::ffff:172.31.37.1:35773    ::ffff:172.31.37.2:9160     ESTABLISHED 8520/java
tcp        0      0 ::ffff:172.31.37.1:53143    ::ffff:172.31.37.3:9160     ESTABLISHED 8520/java
tcp        0      0 ::ffff:172.31.37.1:56411    ::ffff:172.31.37.1:9160     ESTABLISHED 8520/java
```

Reproduced with Cassandra 2.0.5 and 2.0.7 and the following build.sbt libraryDependencies...

``` scala
libraryDependencies ++= Seq(
  "org.apache.spark" %% "spark-core" % "0.9.1" % "provided",
  "com.tuplejump" %% "calliope" % "0.9.0-U1-C2-EA",
  "com.datastax.cassandra" % "cassandra-driver-core" % "2.0.1"
)
```

Not sure if the connections to 9160 are being leaked because of a configuration problem or because of some bug in Calliope or the Cassandra client yet. Please let me know if you have any ideas for things to look for.

**Comment by milliondreams** [comment]
This looks very similar to this one - https://issues.apache.org/jira/browse/CASSANDRA-6169
So ideally should not be produced in 2.0.5 to 2.0.7 :(

Are you suing the same version of C\* in your M/R job and server?

Also can you describe your cluster and check on as to how many tasks is your job creating?

**Comment by brenttheisen** [comment]
Thanks for pointing me in the right direction. The Jira issue you pointed me to describes the same symptom I was seeing except the issue only addressed the input format, not the output format where I was seeing it. After looking around Cassandra's Jira I was able to find this issue that appears to address the output format issue I was seeing... https://issues.apache.org/jira/browse/CASSANDRA-6665

After upgrading my Calliope working copy from cassandra-all 2.0.5 to 2.0.7 the issue no longer occurs. I'm pretty sure that the patch from CASSANDRA-6665 that first shipped in 2.0.6 is the reason it is fixed. I have created a PR to upgrade Calliope's cassandra-all dependency to 2.0.7 in #24 that contains the same code I used to test with.

After upgrading Calliope's cassandra-all dependency to 2.0.7 I found yet another issue with Cassandra's Hadoop implementation that causes CQL3 jobs that do not specify a page size to fail. For that issue I submitted a patch to Cassandra in... https://issues.apache.org/jira/browse/CASSANDRA-7198

Until a Cassandra release ships with a fix for CASSANDRA-7198 the work around to get CQL3 RDDs to work is to always set a page size when using `Cql3CasBuilder`. The default page size in cassandra-all is 1000 so I just changed to...

``` scala
new Cql3CasBuilder("keyspace", "table").setPageSize(1000)
```

After all that I've gotten my Spark job to complete successfully so I'm going to close this issue because none of the causes are Calliope specific.

**Comment by milliondreams** [comment]
@brenttheisen Glad that it all worked for you. Till now we have relied totally on Cassandra's hadoop components for Calliope base, but as we find more and more similar issues and moreover as it is focused too much on Hadoop (there are a few hadoop specific pieces of code in there) we do plan to bring in and maintain that code as part of Calliope. 

This will allow us to take the responsibility of the entire codebase and release independent of C*. Will also help us to optimize the internal guts towards Spark use case and move away from Hadoop!
